src/libexpr/primops.cc

```diff
@@ -1445,14 +1445,19 @@ static RegisterPrimOp primop_toPath({
    corner cases. */
 static void prim_storePath(EvalState & state, const PosIdx pos, Value * * args, Value & v)
 {
-    if (evalSettings.pureEval)
-        state.debugThrowLastTrace(EvalError({
-            .msg = hintfmt("'%s' is not allowed in pure evaluation mode", "builtins.storePath"),
-            .errPos = state.positions[pos]
-        }));
-
     PathSet context;
-    Path path = state.checkSourcePath(state.coerceToPath(pos, *args[0], context, "while evaluating the first argument passed to builtins.storePath"));
+    Path uncheckedPath = state.coerceToPath(pos, *args[0], context, "while evaluating the first argument passed to builtins.storePath");
+
+    if (evalSettings.pureEval) {
+        // TODO feature flag?
+        //     state.debugThrowLastTrace(EvalError({
+        //         .msg = hintfmt("'%s' is not allowed in pure evaluation mode", "builtins.storePath"),
+        //         .errPos = state.positions[pos]
+        // }));'
+        state.allowPath(uncheckedPath);
+    }
```

Comment:
I'll look for a workaround, as I know this is controversial. `builtins.storePath` is referentially transparent, and only undesirable for some. If I can find an alternative for this change, I'll use that instead.